### PR TITLE
Improve Audio Component

### PR DESCRIPTION
**Pull request by hannahblair**
## Description

This PR updates our Audio component. 

to do:

- [x] Visual waveform (Recordings & Uploads)
- [x] Play and pause controls
- [x] Show timestamps
- [x] Playback Speed
- [x] Download file
- [x] Select snippets
- [x] Clear audio
- [x] Recording
- [x] Timestamps while recording
- [x] Pause and resume recording
- [x] Audio device enumeration
- [x] Trim recordings
- [x] Trim uploaded audio
- [x] Streaming
- [x] WASM changes
- [x] Dispatch events  
    - [x] Events.stream
    - [x] Events.change
    - [x] Events.clear
    - [x] Events.play
    - [x] Events.stop
    - [x] Events.pause
    - [x] Events.start_recording
    - [x] Events.stop_recording
    - [x] Events.pause_recording (new - add this to backend)
    - [x] Events.upload
- [x] Tests
- [x] Update backend with new params
- [x] Update documentation
    - [x] New params (waveform settings)
    - [x] Update demos 
- [ ] Allow passing captions
- [ ] Allow passing video

Closes: #(issue)

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

**Comment by abidlabs** [comment]
Really nice UI/UX @hannahblair! 🔥🔥

Here's what I tested and a few issues I found:

1. Testing displaying an audio file: both interactive and static mode work great

```py
with gr.Blocks() as demo:
    gr.Audio("cantina.wav")
    gr.Audio("cantina.wav", interactive=True)
    
demo.launch()
```

Closes: #5299

2. Tested input and output `Interface`: 

```py
import gradio as gr

gr.Interface(lambda x:x, "audio", "audio").launch()
```

Works fantastic. I had one question: it looks like we duplicate the icons in the interactive Audio component: we have two icons to trim the audio, and two to reset. Is this intentional? If not, my suggestion would be to:
* remove the pencil icon altogether
* change the behavior of the "circling back" icon to reset the trim. That's what I would have expected it to do, but it actually reset the entire audio component

3. Tested trimming: the UI is great, and it processing the value well. My main question was what does the number in the orange represent? I thought it was the length of the trimmed audio, but it always was always stuck at 0:01 even when I adjusted the length of the trimmed section

image

Closes: #5320

3. Tested microphone:

```
import gradio as gr

gr.Interface(lambda x:x, "mic", "audio").launch()
```
 
* I wasn't able to get the microphone value to be sent to the backend. You can see in the example above, even if you record, and then click submit, it always sends an empty value
* Its awesome that you now have a dropdown to select the microphone sources 🚀 -- but I noticed that before you grant microphone permissions, you see an empty dropdown which looks unusual:

image

* Love that you can pause and resume recordings. Nit: should we add an icon before resume here? Nit: should we add an icon before the audio sources? 

image

Closes: #4987, Closes: #5720

4. Tested setting min and max length:

```
import gradio as gr

gr.Interface(lambda x:x, gr.Audio(max_length=1, source="upload"), "audio").launch()
```

* Works great -- enforces it on the backend, which is ideal
* I wonder if we should also enforce it on the frontend? Like if someone uploads an audio file that is too long, perhaps they should know immediately so that they can trim it to get it into the right range. 

Closes: #428

Really awesome redesign @hannahblair!

Closes: #467, Closes: #5321

**Comment by abidlabs** [comment]
As discussed in sync, it would be really nice if we could allow source to be a list `["microphone", "upload"]`, which would allow a user to choose either a microphone or upload a file (and close a long-standing issue: #3099). On the other hand, I love how clean this UI is and maybe that would clutter things up too much. Let's see how @pngwn implements it for the Image component and if it makes sense to apply here as well

File: gradio/components/audio.py [hunk 1]
```diff
@@ -102,13 +105,11 @@ def __init__(
...
+            max_length: The maximum length of audio (in seconds) that the user can pass into the prediction function. If None, there is no maximum length.
+            waveform_options: A dictionary of options for the waveform display. Options include: waveform_color (str), waveform_progress_color (str), show_controls (bool). Default is None, which uses the default values for these options.
         """
-        valid_sources = ["upload", "microphone"]
-        source = source if source else ("microphone" if streaming else "upload")
-        if source not in valid_sources:
-            raise ValueError(
-                f"Invalid value for parameter `source`: {source}. Please choose from one of: {valid_sources}"
-            )
+        source = list(set(source)) if source is not None else (["microphone"] if streaming else ["microphone", "upload"])
```
Comment (abidlabs) [inline gradio/components/audio.py:112]: I like that you convert strings to lists, but just wondering, do you need the `set` here? Or could you just do:
```suggestion
        source = list(source) if source is not None else (["microphone"] if streaming else ["microphone", "upload"])
```

Comment (abidlabs) [reply gradio/components/audio.py:112]: Ideally if a user accidentally passes in a string (e.g. "upload"), it should treat it as the same as ["upload"], which `list(source)` does, but `list(set(source))` does not

**Comment by abidlabs** [comment]
Gave this another spin, love the joint upload/mic functionality. It seems that most of my original issues have been addressed, thank you!

> It is indeed the length of the trimmed audio - your audio there is fairly short at 0:03, so your selection duration is rounded to the nearest second and we haven't got any more granularity in that value

I tried with a larger audio clip, and I'm still seeing this behavior. The second duration remains fixed, even if I change the width of the trimming window. You can see here, the trimming window is covering most of the width of the 11 second clip, but its only showing 3 seconds as the duration in orange:

image

The color of the "selected" source could be improved in light mode. Right now it becomes white, which is very hard to see on the white background: 

image

(it looks great in dark mode btw).

Very excited for this 🚀🚀

**Comment by pngwn** [comment]
Everything looks great from my limited testing. The only thing I did notice was that certain things didn't work as expected on mobile. Could you just test the buttons + trimming behaviour on mobile to make sure everything works as expected? We might need to scatter in some touch events.

File: gradio/components/audio.py [hunk 2]
```diff
@@ -57,7 +58,7 @@ def __init__(
         self,
         value: str | Path | tuple[int, np.ndarray] | Callable | None = None,
         *,
-        source: Literal["upload", "microphone"] | None = None,
+        source: list[Literal["upload", "microphone"]] | None = None,
```
Comment (pngwn) [inline gradio/components/audio.py:61]: Could we rename this to `sources` to match the image component?

Comment (abidlabs) [reply gradio/components/audio.py:61]: Let's keep note of this as a deprecation

File: gradio/components/audio.py [hunk 3]
```diff
@@ -102,24 +108,26 @@ def __init__(
...
-            raise ValueError(
-                f"Invalid value for parameter `source`: {source}. Please choose from one of: {valid_sources}"
-            )
-        self.source = source
+        sources = (
+            list(sources)
+            if sources is not None
+            else (["microphone"] if streaming else ["microphone", "upload"])
+        )
+        self.sources = sources
```
Comment (abidlabs) [inline gradio/components/audio.py:120]: nit: slightly more concise
```suggestion
        self.sources = (
            list(sources)
            if sources is not None
            else (["microphone"] if streaming else ["microphone", "upload"])
        )
```

File: gradio/events.py [hunk 4]
```diff
@@ -464,6 +464,10 @@ class Events:
         "start_recording",
         doc="This listener is triggered when the user starts recording with the {{ component }}.",
     )
+    pause_recording = EventListener(
```
Comment (abidlabs) [inline gradio/events.py:467]: Cool addition

File: gradio/components/audio.py [hunk 5]
```diff
@@ -102,24 +108,26 @@ def __init__(
...
-        if source not in valid_sources:
-            raise ValueError(
-                f"Invalid value for parameter `source`: {source}. Please choose from one of: {valid_sources}"
-            )
-        self.source = source
+        sources = (
+            list(sources)
+            if sources is not None
+            else (["microphone"] if streaming else ["microphone", "upload"])
+        )
```
Comment (abidlabs) [inline gradio/components/audio.py:119]: nit: handle the str case as well for better DX
```suggestion
        if sources is None:
            sources = ["microphone"] if streaming else ["microphone", "upload"]
        if isinstance(sources, str):
            sources = [sources]
        elif isinstance(sources, list):
            pass
        else:
            raise ValueError('`sources` must a list consisting of elements in ["microphone", "upload"]')
```

**Review by abidlabs — APPROVED** [review]
LGTM @hannahblair so awesome!

File: gradio/templates.py [hunk 6]
```diff
@@ -385,7 +385,7 @@ def __init__(
         self,
         value: str | tuple[int, np.ndarray] | Callable | None = None,
         *,
-        source: Literal["microphone"] = "microphone",
+        sources: list[Literal["upload", "microphone"]] | None = None,
```
Comment (abidlabs) [inline gradio/templates.py:388]: T
```suggestion
        sources: list[Literal["upload", "microphone"]] | None = ["microphone"],
```

**Pull request merged**